multichain-aggregator/multichain-aggregator-logic/src/repository/address_token_balances.rs

```diff
@@ -49,3 +62,110 @@ where
 
     Ok(())
 }
+
+pub type ListAddressTokensPageToken = (Option<BigDecimal>, BigDecimal, i64);
+
+pub async fn list_by_address<C>(
+    db: &C,
+    address: alloy_primitives::Address,
+    token_type: Option<TokenType>,
+    chain_ids: Vec<i64>,
+    page_size: u64,
+    page_token: Option<ListAddressTokensPageToken>,
+) -> Result<
+    (
+        Vec<ExtendedAddressTokenBalance>,
+        Option<ListAddressTokensPageToken>,
+    ),
+    DbErr,
+>
+where
+    C: ConnectionTrait,
+{
+    let tokens_rel = Entity::belongs_to(tokens::Entity)
+        .from((Column::TokenAddressHash, Column::ChainId))
+        .to((tokens::Column::AddressHash, tokens::Column::ChainId))
+        .into();
+
+    let fiat_balance_iden = Alias::new("fiat_balance");
+    let fiat_balance_col = Expr::col(fiat_balance_iden.clone());
+
+    let balances = Entity::find()
+        .expr_as(fiat_balance_query(), fiat_balance_iden.to_string())
+        .join(JoinType::InnerJoin, tokens_rel)
+        .filter(Column::AddressHash.eq(address.as_slice()))
+        .filter(Column::ChainId.is_in(chain_ids))
+        .filter(Column::Value.gt(0))
+        .apply_if(token_type, |q, token_type| {
+            q.filter(tokens::Column::TokenType.eq(token_type))
+        })
+        .apply_if(page_token, |q, page_token| {
+            let (fiat_value, value, id) = page_token;
+            match fiat_value {
+                None => q.filter(Expr::cust_with_exprs(
+                    "$1 IS NULL AND ($2 < $3 OR ($2 = $3 AND $4 < $5))",
+                    [
+                        fiat_balance_query(),
+                        Column::Value.into_simple_expr(),
+                        Expr::value(value),
+                        Column::Id.into_simple_expr(),
+                        Expr::value(id),
+                    ],
+                )),
+                Some(fiat_value) => q.filter(Expr::cust_with_exprs(
+                    "$1 < $2 OR $1 IS NULL OR ($1 = $2 AND ($3 < $4 OR ($3 = $4 AND $5 < $6)))",
+                    [
+                        fiat_balance_query(),
+                        Expr::value(fiat_value),
+                        Column::Value.into_simple_expr(),
+                        Expr::value(value),
+                        Column::Id.into_simple_expr(),
+                        Expr::value(id),
+                    ],
+                )),
+            }
+        })
+        .order_by_with_nulls(fiat_balance_col, Order::Desc, NullOrdering::Last)
+        .order_by_desc(Column::Value)
+        .order_by_desc(Column::Id)
+        .limit(page_size + 1)
```

Comment:
Are you sure the pagination implementation is correct? It definitely differs from the standard Blockscout pagination.

In all API endpoints, the user cannot rely on the number of items returned by the endpoint based on any query parameter. You're using the `page_size` query parameter to request a specific number of items to return.

Commonly, in Blockscout APIs, the `items_count` parameter is introduced to help track the total number of items on the UI side. You can set any `items_count` value in the input request, and in the pagination dictionary, you'll get `items_count = input_items_count + fixed_page_size`.

Are there any reasons for implementing it this way, differently from the pagination used in other APIs?

Reply:
`items_count` in the pagination of Blockscout REST API, as it turned out, is unused by API consumers. We scheduled to remove it in v10 of the Elixir backend. The corresponding task for the reference https://github.com/blockscout/blockscout/issues/12901.

Reply:
@EvgenKor I changed the pagination a little to encode page token as base64-encoded JSON array. This solves a lot of potential issues when one of the page token parts is of `String` type and it could contain commas (we used commas to split original page token string previously).